### ISSUE TITLE
false positive for no-unused-vars argsIgnorePattern

### ISSUE THREAD

**Issue by Fandekasp** [eslint/eslint#11139]
**Tell us about your environment**

* **ESLint Version:** 5.9.0
* **Node Version:** 10.12.0
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
extends:
  - eslint-config-with-prettier
  - plugin:redux-saga/recommended
  - plugin:security/recommended

settings:
  react:
    version: 16.6.0
  import/resolver:
    babel-plugin-root-import: {}
    typescript: {} # use tsconfig.json
  import/parsers:
    typescript-eslint-parser: [ .ts, .tsx ]


parser: babel-eslint

plugins:
  - immutable
  - import
  - material-ui
  - redux-saga
  - security
  - typescript

parserOptions:
  ecmaVersion: 2018
  sourceType: "module"
  allowImportExportEverywhere: false
  codeFrame: false
  ecmaFeatures:
    legacyDecorators: true
    jsx: true

globals:
  firebase: true
  requirejs: true
  window: true
  _: true
  expect: true
  jest: true
  # enzyme vars (check jestsetup.js)
  shallow: true
  render: true
  mount: true

env:
  amd: true # for require.js (request() and define())
  browser: true
  commonjs: true
  es6: true
  mocha: true

rules:
  class-methods-use-this: 2
  flowtype/no-types-missing-file-annotation: 0
  immutable/no-mutation: 2
  import/named: 2
  import/order: error
  import/no-extraneous-dependencies: 0
  import/no-unresolved: error
  jsx-a11y/anchor-is-valid:
  - error
  - components:
    - Link
    specialLink:
    - to
  no-restricted-syntax:
    - error
    - selector: 'ForInStatement'
      message: 'for..in loops iterate over the entire prototype chain, which is virtually never what you want. Use Object.{keys,values,entries}, and iterate over the resulting array.'
    - selector: 'LabeledStatement'
      message: 'Labels are a form of GOTO; using them makes code confusing and hard to maintain and understand.'
    - selector: 'WithStatement'
      message: '`with` is disallowed in strict mode because it makes code impossible to predict and optimize.'
  no-else-return:
    - "error"
    - allowElseIf: true
  no-unused-vars: ["error", { "argsIgnorePattern": "^_" }]
  one-var:
    - error
    - var: "always"
  prefer-destructuring:
    - error
    - object: false
  react/jsx-filename-extension: 0
  react/forbid-foreign-prop-types: 2
  # We'll delete proptypes after finishing typescript integration
  react/prop-types: [2, {skipUndeclared: 1}]
  react/sort-prop-types: "error"
  sort-imports: 0
  sort-keys: 0
  sort-vars: 2
  strict: 0
  typescript/class-name-casing: "error"
  typescript/no-unused-vars: "error"
  typescript/type-annotation-spacing: "error"

overrides:
  - files: ["*.ts", "*.tsx"]
    parser: "typescript-eslint-parser"
    parserOptions:
        jsx: true

# vi: ft=yaml
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```es6
const selectSponsors = createSelector(
  selectHosts,
  ({hosts}) => {
    return Object.keys(hosts)
      .slice(12)
      .map(key => {
        const {pk, slug, name, categoryName, imageUrl, ..._rest} = hosts[key];
        return {pk, slug, name, categoryName, imageUrl};
      });
  },
);
```

```bash
$ eslint \"src/**\" --cache 2>/dev/null
```

**What did you expect to happen?**
No linting error on unused property _rest

**What actually happened? Please include the actual, raw output from ESLint.**
```
  15:59  error    '_rest' is assigned a value but never used                                                             no-unused-vars 
```

Note: The error disappears with the following change
```diff
-   no-unused-vars: ["error", { "argsIgnorePattern": "^_" }]
+   no-unused-vars: ["error", { "argsIgnorePattern": "^\.\.\._" }]
```
This means that eslint assumes the destructuring operator `...` is part of the variable name, which it shouldn't.

**Comment by not-an-aardvark** [comment]
Hi, thanks for the report. However, this is working as intended; the `argsIgnorePattern` option only applies to function parameters. You can use the `varsIgnorePattern` option to ignore specific variables.

You might also consider using the [`ignoreRestSiblings`](https://eslint.org/docs/rules/no-unused-vars#ignorerestsiblings) option to ignore variables like the one in your code.

**Comment by Fandekasp** [comment]
@not-an-aardvark in our case, we want to ignore rest, not its siblings.

> the argsIgnorePattern option only applies to function parameters

Got it. Still, the no-unused-vars rule covers much more than function parameters, so there's something missing here :)

**Comment by not-an-aardvark** [comment]
>  in our case, we want to ignore rest, not its siblings.

You're right, I missed that. In that case, why not just omit the variable?

```diff
 const selectSponsors = createSelector(
   selectHosts,
   ({hosts}) => {
     return Object.keys(hosts)
       .slice(12)
       .map(key => {
-        const {pk, slug, name, categoryName, imageUrl, ..._rest} = hosts[key];
+        const {pk, slug, name, categoryName, imageUrl} = hosts[key];
         return {pk, slug, name, categoryName, imageUrl};
       });
   },
 );
```

> Still, the no-unused-vars rule covers much more than function parameters, so there's something missing here :)

I'm not sure I understand -- you can use the `varsIgnorePattern` option to cover the cases that `argsIgnorePattern` doesn't cover, and the combination of the two options should allow you to ignore everything covered by the no-unused-vars rule.

**Comment by Fandekasp** [comment]
@not-an-aardvark 	oh. Sorry, you're totally right ^^
Moreover, the `"^\.\.\._"` pattern actually broke my config, it didn't get it to pass.

Closing the issue, thanks again!
